clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -215,7 +215,27 @@ protected void onJoinComplete(int generation,
 
         // update partition assignment
         subscriptions.assignFromSubscribed(assignment.partitions());
-        this.joinedSubscription = subscriptions.subscription();
+
+        // check if the assignment contains some topics that is not in the original
+        // subscription, if yes we will obey what leader has decided and add these topics
+        // into the subscriptions
+        //
+        // TODO this part of the logic should be removed once we allow regex on leader assign and remove joinedSubscription
+        Set<String> addedTopics = new HashSet<>();
+        for (TopicPartition tp : subscriptions.assignedPartitions()) {
+            if (!subscriptions.subscription().contains(tp.topic()))
```

Comment:
I'm puzzling a bit over whether we should be using `joinedSubscription` here instead of `subscriptions.subscription()`, or whether it matters. 

Seems it should be `joinedSubscription`. Suppose that `joinedSubscription` contained only [A] at the time of rebalance. Topic B is then created by the leader and assigned. Before the rebalance completes, the consumer discovers topic B, and `subscription` is updated to [A, B], while `joinedSubscription` is still [A]. The consumer then receives the assignment containing partitions from both A and B, but `addedTopics` will be empty (since B was already added to `subscription`). The consumer will then notice that `joinedSubscription` is [A], while `subscription` is [A, B], and request an unneeded rebalance.